src/transactions/TransactionFrame.h

```diff
@@ -240,68 +193,82 @@ class TransactionFrame : public TransactionFrameBase
-    bool checkSorobanResourceAndSetError(Application& app,
-                                         uint32_t ledgerVersion) override;
+                                 AccountID const& accountID) const;
+    bool checkExtraSigners(SignatureChecker& signatureChecker) const;
+
+    std::pair<bool, TransactionResultPayloadPtr>
+    checkValidWithOptionallyChargedFee(
+        Application& app, AbstractLedgerTxn& ltxOuter, SequenceNumber current,
+        bool chargeFee, uint64_t lowerBoundCloseTimeOffset,
+        uint64_t upperBoundCloseTimeOffset) const;
+    std::pair<bool, TransactionResultPayloadPtr>
+    checkValid(Application& app, AbstractLedgerTxn& ltxOuter,
+               SequenceNumber current, uint64_t lowerBoundCloseTimeOffset,
+               uint64_t upperBoundCloseTimeOffset) const override;
+    bool checkSorobanResourceAndSetError(
+        Application& app, uint32_t ledgerVersion,
+        TransactionResultPayloadPtr resPayload) const override;
+
+    TransactionResultPayloadPtr createResultPayload() const override;
+
+    TransactionResultPayloadPtr
+    createResultPayloadWithFeeCharged(LedgerHeader const& header,
+                                      std::optional<int64_t> baseFee,
+                                      bool applying) const override;
 
     void
     insertKeysForFeeProcessing(UnorderedSet<LedgerKey>& keys) const override;
     void insertKeysForTxApply(UnorderedSet<LedgerKey>& keys) const override;
 
     // collect fee, consume sequence number
-    void processFeeSeqNum(AbstractLedgerTxn& ltx,
-                          std::optional<int64_t> baseFee) override;
+    TransactionResultPayloadPtr
+    processFeeSeqNum(AbstractLedgerTxn& ltx,
+                     std::optional<int64_t> baseFee) const override;
 
     // apply this transaction to the current ledger
     // returns true if successfully applied
     bool apply(Application& app, AbstractLedgerTxn& ltx,
-               TransactionMetaFrame& meta, bool chargeFee,
-               Hash const& sorobanBasePrngSeed);
+               TransactionMetaFrame& meta,
+               TransactionResultPayloadPtr resPayload, bool chargeFee,
+               Hash const& sorobanBasePrngSeed) const;
     bool apply(Application& app, AbstractLedgerTxn& ltx,
                TransactionMetaFrame& meta,
-               Hash const& sorobanBasePrngSeed = Hash{}) override;
+               TransactionResultPayloadPtr resPayload,
+               Hash const& sorobanBasePrngSeed = Hash{}) const override;
 
     // Performs the necessary post-apply transaction processing.
     // This has to be called after both `processFeeSeqNum` and
     // `apply` have been called.
     // Currently this only takes care of Soroban fee refunds.
-    void processPostApply(Application& app, AbstractLedgerTxn& ltx,
-                          TransactionMetaFrame& meta) override;
+    void
+    processPostApply(Application& app, AbstractLedgerTxn& ltx,
```

Comment:
We need to take `TransactionResultPayload` here, and a pointer is more convenient from a testing perspective. We want the internal `TransactionResultPayload` in `TransactionTestFrame` to be correct whether it is called via the test only functions that don't include an explicit `TransactionResultPayload` or the normal non-test functions. To do this we always update the internal result pointer to whatever parameter is passed in via these functions.

In a previous iteration, I had all these parameters as references, but then I had to implement `shared_from_this` to update the testing pointer. I think pointers make sense here over references, just because the canonical `TransactionResultPayload` type that callers will have access to is always a `TransactionResultPayloadPtr`, since the constructor is private and the factory function returns a pointer type.

Reply:
I'm not quite sold on designing the interfaces in certain ways just to cater to the test-only utilities. But I think we can also revisit this later, as it's not a big deal overall.